User Profile Class

Mariana-Ayup opened the pull request:
Correction of class UserProfile, UserProfileService, UserProfileEndpoint

MariAyup committed:
Merge branch 'master' of https://github.com/elismer/ProgramRepairGolf-DOSE2019

MariAyup committed:
Merge branch 'master' of github.com:Mariana-Ayup/ProgramRepairGolf-DOSE2019

MariAyup committed:
Merge branch 'master' of git@github.com:saiema/ProgramRepairGolf-DOSE2019.git

MariAyup committed:
Creacion de tabla profile_user

MariAyup committed:
Correccion de error

MariAyup committed:
Merge branch 'master' of git@github.com:saiema/ProgramRepairGolf-DOSE2019.git

MariAyup committed:
Commit clase profile

MariAyup committed:
Clase user profile

MariAyup committed:
Tabla User Profile

MariAyup committed:
User Profile Class

MariAyup committed:
Merge branch 'master' of git@github.com:saiema/ProgramRepairGolf-DOSE2019.git

MariAyup committed:
Merge branch 'master' of git@github.com:saiema/ProgramRepairGolf-DOSE2019.git

MariAyup committed:
Merge branch 'master' of git@github.com:saiema/ProgramRepairGolf-DOSE2019.git

MariAyup committed:
Merge branch 'master' of git@github.com:saiema/ProgramRepairGolf-DOSE2019.git

MariAyup committed:
Merge branch 'master' of github.com:saiema/ProgramRepairGolf-DOSE2019

MariAyup committed:
Corrections Tabla users_profile

MariAyup committed:
Correction Class App

MariAyup committed:
Correction Class User

MariAyup committed:
Corrections Class UserProfile

MariAyup committed:
Class UserProfileEndpoint

MariAyup committed:
Class UserProfileService

MariAyup committed:
Correction Class UserProfileTest

MariAyup committed:
Corrections class UserProfile, User,UserProfileEndpoint y
UserProfileService

saiema committed:
Delete -X

This file is just a http request result, it is not relevant

saiema commented on Profile.java:
```diff
@@ -0,0 +1 @@
+Class Profile
```
The class is missing, it is just the first line of the file

saiema commented on src/main/java/unrc/dose/App.java:
```diff
@@ -45,7 +45,7 @@ public static void main( String[] args ) {
 										urlReset = "";
 									}
 								}
-								if (request.requestMethod() != "OPTIONS"){
```
This line must be restored: the line should be `if (request.requestMethod() != "OPTIONS"){`

saiema committed:
Delete Profile.java

This is just a junk file that wasn't deleted, the Users Profile is implemented in class UserProfile

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,160 @@
+ * Table name: user_profiles (
+  id 	       integer NOT NULL AUTO_INCREMENT,
+  user_id      integer,
+  displayName  varchar(50),
```
The column name doesn't follow conventions

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,160 @@
+    	}
+    	
+    	
+    }*/
```
Dead code should be removed

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,108 @@
+          					if (!User.exists(Integer.parseInt(req.params("id")))) {
+           					throw new UserNotFoundException(String.valueOf(Integer.parseInt(req.params("id"))));
+          					}
+          					UserProfile up = (UserProfile) UserProfile.find("user_id = ?", req.params("id")).get(0);
```
There should be a method in `UserProfileService` that encapsulate this search.

A method like `UserProfileService#findById(int) : UserProfile` (maybe even using `Optional<UserProfile>`)

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,108 @@
+          				    up.setDisplayName (req.queryParams("displayName"));
+                            up.setTwitterId (req.queryParams("twitterId"));
+                            up.saveIt();
+                            return up.toJson(true, "user_id", "displayName","twitterId");
```
`UserProfile `does not seem to have a `toJson` method

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,108 @@
+                            .withPathParam()
+                            	.withName("twitterId")
+                            	.withDescription("user profile twitter id").and()
+                            .withResponseType(String.class),
```
The get should return a `UserProfile` (as JSON), and as such it should not have any parameters

saiema commented on src/main/java/unrc/dose/UserProfileService.java:
```diff
@@ -0,0 +1,17 @@
+      
+      return UserProfile.createUserProfile(userId, displayName, twitterId).toJson(true);
+
+  }
```
The method should be renamed as `create` and `UserProfile` does not seem to have a `toJson` method

saiema commented on src/migrations/20191104165250_create_profile_table.sql:
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS user_profiles (
+  id 	       integer NOT NULL AUTO_INCREMENT,
+  user_id      integer,
+  displayName  varchar(50),
```
The column name does not follow naming conventions

saiema commented on src/test/java/unrc/dose/InsertTest.java:
```diff
@@ -0,0 +1,33 @@
+import org.junit.Test;
+
+
+public class InsertTest {
```
This class was used as a way to insert a new user, it should be removed from the Pull Request

saiema commented on src/test/java/unrc/dose/UserProfileTest.java:
```diff
@@ -0,0 +1,72 @@
+
+public class UserProfileTest {
+	
+	private User newUser = User.set("Enzo" , "Ferrari" , "F40@gmail.com" , false);
```
This should be inside a `@BeforeClass` method

saiema commented on src/test/java/unrc/dose/UserProfileTest.java:
```diff
@@ -0,0 +1,72 @@
+		
+		UserProfile up = UserProfile.createUserProfile(newUser.getId(),"Enzo","@Enzo");
+		assertEquals(true,up.getDisplayName());
+		assertEquals(true,up.getTwitterId());
```
The types are incorrect, the getter methods from `UserProfile` return a `String`

saiema reviewed (changes requested):
The only thing that does not requiere to be changed (although there is a correction for it) is the displayName column name

MariAyup committed:
Correction class App

MariAyup committed:
Delete InsertTest

MariAyup committed:
Correction UserProfileTest

MariAyup committed:
Merge remote-tracking branch 'origin/master'

MariAyup committed:
Corrections

MariAyup committed:
Corrections tests y user

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,120 @@
+    	set(TWITTER_ID,twitterId);
+    	
+    }
+    public static String toJson(List<UserProfile> list){
```
The toJson method should not be static and should take no parameters, since it will be applied to `this` instance

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,120 @@
+    public static String toJson(List<UserProfile> list){
+        String aux="[";
+        for(int i=0;i<list.size();i++){
+          aux=aux+list.get(i).toJson(true);
```
The `toJson(boolean)` method is still not found in this class, does this method comes from `Model`?

saiema commented on src/main/java/unrc/dose/UserProfileService.java:
```diff
@@ -0,0 +1,44 @@
+			throw new UserNotFoundException(String.valueOf(userId));
+			}
+	     UserProfile up = UserProfile.findFirst("user_id = ?", userId);
+	     if (up.getDisplayName().isEmpty() || up.getTwitterId().isEmpty()) {
```
This code is checking for a profile with no displayName and no twitterId (empty values) but the first time a user is created, it has an UserProfile with this specific characteristics, so no profile could ever be modified.

The check should check if a UserProfile exists for a specific user, not check the values it has

saiema commented on src/test/java/unrc/dose/UserProfileTest.java:
```diff
@@ -0,0 +1,75 @@
+			 UserProfile.createUserProfile(newUser.getId(),"Enzo","MLEnzo");
+			 fail("expected IllegalArgumentException");
+			  } catch(IllegalArgumentException e) {}
+	}
```
The same User is created on all tests, the user should be created once in the beforeAll method

saiema reviewed (changes requested):
There are still issues with this PR

MariAyup committed:
Corrections

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,117 @@
+   			}
+   	 return true;
+   	 
+    }
```
This method should be implemented in `UserProfileService`. The method `UserProfile#exists(int)` returns if a `UserProfile` with a particular id exists, but it searches by the profile's id, this is using the `userId` which are not the same.

The method must return a boolean if a profile exists or not, but the only return value in this method is `true`.

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,93 @@
+                            .withResponseType(String.class),
+                        (req, res) -> {
+                        	
+                        	return UserProfileService.searchById(Integer.parseInt(req.params("id")));
```
The returned value should be a JSON representation of a UserProfile, but the methods name suggests that this only returns a UserProfile (if found, what happens if there is no UserProfile or User?)

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,93 @@
+                    (req, res) -> {
+                    	
+                    	    return UserProfileService.updateUp(Integer.parseInt(req.params("id")), 
+                    	   		req.queryParams("displayName"),req.queryParams("twitterId"));
```
What happens if the User doesn't exists?

saiema commented on src/main/java/unrc/dose/UserProfileService.java:
```diff
@@ -0,0 +1,42 @@
+  /**
+  *Invokes the method which creates a user profile and transforms to json.
+  **/
+  public static String userprofile(Integer userId, String displayName, String twitterId) {
```
this method's name doesn't convey what the method does, you should change the name to something like createProfile

saiema commented on src/main/java/unrc/dose/UserProfileService.java:
```diff
@@ -0,0 +1,42 @@
+          }
+      	UserProfile up = UserProfile.findFirst("user_id = ?", userId); 
+      	return up.toJson(true, "user_id", "displayName","twitter_id");       	
+  }
```
This method should return a `UserProfile`, a check to see if a profile exists for a particular userId should be either a precondition of this method or a part of this method's code.

saiema reviewed (changes requested):
the code is very inconsistent in UserEndpoint, either all get, post, and put methods should check that the User exists for a particular userId or this check should be done in the UserService method, now is a mix of both approaches.

The previous corrections have not been resolved

MariAyup committed:
Corrections methods of UserProfile, UserProfileService and
UserProfileEndpoint

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,125 @@
+    	if (displayName == null) {
+			throw new IllegalArgumentException("The user's display name cannot be null");
+		} 
+    	if (!displayName.isEmpty()&& (!displayName.matches("[a-zA-Z ]{2,254}"))){
```
This regular expression allows the displayName to be all blank spaces

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,125 @@
+    	  if (!User.exists(userId)){
+              throw new UserNotFoundException(String.valueOf(userId));
+              }
+    	  UserProfile up = UserProfile.findById(userId);
```
The `findById` method is meant to search a `UserProfile` by `id`, but the argument used is a `userId`. This will only work by coincidence when a `UserProfile` has the same `id` as the `User` to which is related.

saiema commented on src/main/java/unrc/dose/UserProfile.java:
```diff
@@ -0,0 +1,125 @@
+    	  if (up == null) {
+    	    return null;
+    	  }
+    	  return up;
```
This code is equivalent to `return up`

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+           					throw new UserNotFoundException(String.valueOf(Integer.parseInt(req.params("id"))));
+          					}
+                    	
+                        return UserProfileService.createProfile(
+                        Integer.parseInt(req.params("id")),req.queryParams("displayName"),
+                        req.queryParams("twitterId"));
```
This method does too many things, it creates a new profile and returns a json representation of it.

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+                     
+                    (req, res) -> {
+                    	
+                    	    return UserProfileService.updateProfile(Integer.parseInt(req.params("id")), 
+                    	   		req.queryParams("displayName"),req.queryParams("twitterId"));
+          					
+                            }
```
There is no checking done, like in the post case

saiema commented on src/main/java/unrc/dose/UserProfileEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+                            .withResponseType(String.class),
+                        (req, res) -> {
+                        	
+                        	return UserProfileService.getProfile(Integer.parseInt(req.params("id")));
+                        	
+                        	}
```
There is no checking done, like in the post case

saiema commented on src/main/java/unrc/dose/UserProfileService.java:
```diff
@@ -0,0 +1,48 @@
+  **/
+  public static String createProfile(Integer userId, String displayName, String twitterId) {
+	  
+      return UserProfile.createUserProfile(userId, displayName, twitterId).toJson(true);
```
No check is done with regards to the userId and if the UserProfile already exists

saiema commented on src/main/java/unrc/dose/UserProfileService.java:
```diff
@@ -0,0 +1,48 @@
+  public static String getProfile(int userId) {
+	  
+		  UserProfile up = UserProfile.findProfile(userId);
+		  return up.toJson(true, "user_id", "displayName","twitter_id");
```
No check is done: the userId is not checked to see if a User exists with that id; and no check is done to see if there is a UserProfile for the specific userId

saiema commented on src/main/java/unrc/dose/UserProfileService.java:
```diff
@@ -0,0 +1,48 @@
+	up.saveIt();
+	return up.toJson(true, "user_id", "displayName", "twitter_id");   
+        
+ }
```
No check is done: the userId is not checked to see if a User exists with that id; and no check is done to see if there is a UserProfile for the specific userId

The method does too many things: it updates a UserProfile and returns a json representation of it.

saiema reviewed (changes requested):
Except from the tests, the rest of the code still have the same issues as before

MariAyup committed:
correction method findProfile

MariAyup committed:
Correction

MariAyup committed:
Correction method findProfile

MariAyup committed:
Correction findProfile

MariAyup committed:
User existence and user profile check

MariAyup committed:
Correction display name
